modules/renter/downloadheap.go

```diff
@@ -0,0 +1,209 @@
+	// by chunkIndex such that the earlier chunks are selected first from the
+	// heap.
+	return dch[i].staticChunkIndex < dch[j].staticChunkIndex
+}
+func (dch downloadChunkHeap) Swap(i, j int)       { dch[i], dch[j] = dch[j], dch[i] }
+func (dch *downloadChunkHeap) Push(x interface{}) { *dch = append(*dch, x.(*unfinishedDownloadChunk)) }
+func (dch *downloadChunkHeap) Pop() interface{} {
+	old := *dch
+	n := len(old)
+	x := old[n-1]
+	*dch = old[0 : n-1]
+	return x
+}
+
+// acquireMemoryForDownloadChunk will block until memory is available for the
+// chunk to be downloaded. 'false' will be returned if the renter shuts down
+// before memory can be acquired.
+func (r *Renter) managedAcquireMemoryForDownloadChunk(udc *unfinishedDownloadChunk) bool {
+	// If the chunk does not need memory, can return immediately with a success
+	// condition - all requried memory has been acquired.
+	if !udc.staticNeedsMemory {
+		return true
+	}
+
+	// The amount of memory required is equal minimum number of pieces plus the
+	// overdrive amount.
+	//
+	// TODO: This allocation assumes that the erasure coding does not need extra
+	// memory to decode a bunch of pieces. Optimized erasure coding will not
+	// need extra memory to decode a bunch of pieces, though I do not believe
+	// our erasure coding has been optimized around this yet, so we may actually
+	// go over the memory limits when we decode pieces.
+	memoryRequired := uint64(udc.staticOverdrive+udc.erasureCode.MinPieces()) * udc.staticPieceSize
+	udc.memoryAllocated = memoryRequired
+	return r.memoryManager.Request(memoryRequired)
+}
+
+// managedAddChunkToDownloadHeap will add a chunk to the download heap in a
+// thread-safe way.
+func (r *Renter) managedAddChunkToDownloadHeap(udc *unfinishedDownloadChunk) {
+	// The purpose of the chunk heap is to block work from happening until there
+	// is enough memory available to send off the work. If the chunk does not
+	// need any memory to be allocated, it should be given to the workers
+	// directly and immediately. This is actually a requirement in our memory
+	// model. If a download chunk does not need memory, that means that the
+	// memory has already been allocated and will actually be blocking new
+	// memory from being allocated until the download is complete. If the job is
+	// put in the heap and ends up behind a job which get stuck allocating
+	// memory, you get a deadlock.
+	if !udc.staticNeedsMemory {
+		r.managedDistributeDownloadChunkToWorkers(udc)
+		return
+	}
+
+	// Put the chunk into the chunk heap.
+	r.downloadHeapMu.Lock()
+	r.downloadHeap.Push(udc)
+	r.downloadHeapMu.Unlock()
+}
+
+// managedBlockUntilOnline will block until the renter is online. The renter
+// will appropriately handle incoming download requests and stop signals while
+// waiting.
+func (r *Renter) managedBlockUntilOnline() bool {
+	for !r.g.Online() {
+		select {
+		case <-r.tg.StopChan():
+			return false
+		case <-time.After(offlineCheckFrequency):
```

Comment:
careful -- might leak a timer, depending on how frequently `managedBlockUntilOnline` is called

Reply:
Wait what? I thought `time.After` was a pretty safe function.

Reply:
How do you leak a timer? I also think this function is only called within a single thread.

Reply:
`time.After` creates a timer that only gets GC'd once the duration elapses. So it "leaks" if you're creating timers faster than they're expiring (on average).

In this case, where the only other `case` is a shutdown signal, I think it's okay. It's only a problem in code like:
```go
for {
    success := make(chan struct{})
    someAsyncFunction(success)
    select {
    case <-success:
        continue
    case <-time.After(longTimeoutDuration):
        return errors.New("timed out")
    }
}
```
because then there's a good chance that you'll be creating new `time.After`s faster than the old `time.After`s expire.